File: Python/find-all-duplicates-in-an-array.py
# Time:  O(n)
# Space: O(1)

# Given an array of integers, 1 <= a[i] <= n (n = size of array),
# some elements appear twice and others appear once.
# Find all the elements that appear twice in this array.
# Could you do it without extra space and in O(n) runtime?
#
# Example:
# Input
#
# [4,3,2,7,8,2,3,1]
#
# Output
#
# [2,3]

class Solution(object):
    def findDuplicates(self, nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """
        result = []
        i = 0
        while i < len(nums):
            if nums[i] != nums[nums[i]-1]:
                nums[nums[i]-1], nums[i] = nums[i], nums[nums[i]-1] 
            else:
                i += 1

        for i in xrange(len(nums)):
            if i != nums[i]-1:
                result.append(nums[i])
        return result

File: C++/find-all-duplicates-in-an-array.cpp
// Time:  O(n)
// Space: O(1)

class Solution {
public:
    vector<int> findDuplicates(vector<int>& nums) {
        vector<int> result;
        for (const auto& i : nums) {
            if (nums[abs(i) - 1] < 0) {
                result.emplace_back(abs(i));
            } else {
                nums[abs(i) - 1] *= -1;
            }
        }
        return result;
    }
};

class Solution2 {
public:
    vector<int> findDuplicates(vector<int>& nums) {
        vector<int> result;
        int i = 0;
        while (i < nums.size()) {
            if (nums[i] != nums[nums[i] - 1]) {
                swap(nums[i], nums[nums[i] - 1]);
            } else {
                ++i;
            }
        }
        for (i = 0; i < nums.size(); ++i) {
            if (i != nums[i] - 1) {
                result.emplace_back(nums[i]);
            }
        }
        return result;
    }
};
